### PR TITLE
Wrap all opaque types as interfaces

### PR DESCRIPTION
**Pull request by klueska**

### CODE REVIEW

File: pkg/nvml/lib.go [hunk 1]
```diff
@@ -149,9 +149,9 @@ var nvmlDeviceGetGridLicensableFeatures = nvmlDeviceGetGridLicensableFeatures_v1
 var nvmlEventSetWait = nvmlEventSetWait_v1
 var nvmlDeviceGetAttributes = nvmlDeviceGetAttributes_v1
 var nvmlComputeInstanceGetInfo = nvmlComputeInstanceGetInfo_v1
-var DeviceGetComputeRunningProcesses = deviceGetComputeRunningProcesses_v1
-var DeviceGetGraphicsRunningProcesses = deviceGetGraphicsRunningProcesses_v1
-var DeviceGetMPSComputeRunningProcesses = deviceGetMPSComputeRunningProcesses_v1
+var deviceGetComputeRunningProcesses = deviceGetComputeRunningProcesses_v1
```
Comment (elezar) [inline pkg/nvml/lib.go:152]: Maybe a silly question: Can you remind me why we don't have an `nvml` prefix here?

File: pkg/nvml/api.go [hunk 2]
```diff
@@ -349,6 +349,210 @@ type Interface interface {
 	GetVgpuDriverCapabilities(Capability VgpuDriverCapability) (bool, Return)
 }
 
+type Device interface {
+	GetName() (string, Return)
```
Comment (elezar) [inline pkg/nvml/api.go:353]: Can we sort these? (even if it is as a follow-up?)

File: pkg/nvml/gpm.go [hunk 3]
```diff
@@ -44,23 +44,23 @@ func (l *library) GpmSampleAlloc(GpmSample *GpmSample) Return {
...
+	return nvmlGpmSampleGet(Device, GpmSample)
 }
 
 // nvml.GpmQueryDeviceSupport()
 type GpmSupportV struct {
-	device Device
+	device nvmlDevice
 }
 
 func (l *library) GpmQueryDeviceSupportV(Device Device) GpmSupportV {
```
Comment (elezar) [inline pkg/nvml/gpm.go:69]: This is probably out of scope, but does the `GpmQueryDeviceSupportV` top-level function (and others like it) add any value? Should we not *always* use the `GpmQueryDeviceSupportV` member function?

File: pkg/nvml/api.go [hunk 4]
```diff
@@ -553,6 +553,11 @@ type Device interface {
 	VgpuTypeGetMaxInstances(VgpuTypeId VgpuTypeId) (int, Return)
 }
 
+type EventSet interface {
+	Wait(Timeoutms uint32) (EventData, Return)
```
Comment (elezar) [inline pkg/nvml/api.go:575]: `Timeoutms` -> `timeoutms`? (or just keep the types).

The case of the local variables is one of the biggest issues with enabling the linter in this package.

File: pkg/nvml/api.go [hunk 5]
```diff
@@ -279,7 +279,7 @@ type Interface interface {
 	DeviceSetVgpuSchedulerState(Device Device, PSchedulerState *VgpuSchedulerSetState) Return
 	DeviceGetVgpuSchedulerCapabilities(Device Device) (VgpuSchedulerCapabilities, Return)
 	GpuInstanceGetComputeInstancePossiblePlacements(GpuInstance GpuInstance, Info *ComputeInstanceProfileInfo) ([]ComputeInstancePlacement, Return)
-	GpuInstanceCreateComputeInstanceWithPlacement(GpuInstance GpuInstance, Info *ComputeInstanceProfileInfo, Placement *ComputeInstancePlacement, ComputeInstance *ComputeInstance) Return
+	GpuInstanceCreateComputeInstanceWithPlacement(GpuInstance GpuInstance, Info *ComputeInstanceProfileInfo, Placement *ComputeInstancePlacement) (ComputeInstance, Return)
```
Comment (elezar) [inline pkg/nvml/api.go:282]: Is this a bug? Does it make sense to do this in a separate PR so as to not have it go "missing" in the changelog?

File: pkg/nvml/api.go [hunk 6]
```diff
@@ -292,7 +292,7 @@ type Interface interface {
 	GpmMetricsGetV(MetricsGet *GpmMetricsGetType) GpmMetricsGetVType
 	GpmMetricsGet(MetricsGet *GpmMetricsGetType) Return
 	GpmSampleFree(GpmSample GpmSample) Return
-	GpmSampleAlloc(GpmSample *GpmSample) Return
+	GpmSampleAlloc() (GpmSample, Return)
```
Comment (elezar) [inline pkg/nvml/api.go:295]: Does it make sense to move this to a separate PR to call it out properly?

File: pkg/nvml/api.go [hunk 7]
```diff
@@ -576,6 +576,12 @@ type EventSet interface {
 	Free() Return
 }
 
+type GpmSample interface {
+	Free() Return
+	Get(Device Device) Return
+	MigGet(Device Device, GpuInstanceId int) Return
```
Comment (elezar) [inline pkg/nvml/api.go:582]: Nit:
* `Device Device` -> `Device` or `device Device`
* `GpuInstanceId int` -> `int` or `gpuInstanceID int`.

**Review by elezar — COMMENTED** [review]
Initial change look good. The main question I have is whether to top-level library functions that operate ON other types make sense, or whether we should always use the functions off the types?

File: pkg/nvml/lib.go [hunk 8]
```diff
@@ -44,14 +44,37 @@ type library struct {
...
+	return newLibrary(opts...)
+}
+
+func newLibrary(opts ...LibraryOption) *library {
+	l := &library{}
+	l.init(opts...)
+	return l
+}
+
+func (l *library) init(opts ...LibraryOption) {
```
[the remark this answers is not in the document]
Comment (klueska) [reply pkg/nvml/lib.go:71]: It's an internal type and only called from two places. One place holds the lock -- the other is when it is frist created so doesnt need to.

File: pkg/nvml/api.go [hunk 9]
```diff
@@ -54,7 +49,597 @@ func SetLibraryOptions(opts ...LibraryOption) error {
...
+	GetVgpuCompatibility(nvmlVgpuMetadata *nvmlVgpuMetadata, PgpuMetadata *nvmlVgpuPgpuMetadata) (VgpuPgpuCompatibility, Return)
+	GetVgpuVersion() (VgpuVersion, VgpuVersion, Return)
+	SetVgpuVersion(VgpuVersion *VgpuVersion) Return
+	VgpuInstanceClearAccountingPids(VgpuInstance VgpuInstance) Return
+	VgpuInstanceGetMdevUUID(VgpuInstance VgpuInstance) (string, Return)
+	VgpuTypeGetCapabilities(VgpuTypeId VgpuTypeId, Capability VgpuCapability) (bool, Return)
+	GetVgpuDriverCapabilities(Capability VgpuDriverCapability) (bool, Return)
 }
+
+// Define package level methods as aliases to Interface methods of libnvml
```
[the remark this answers is not in the document]
Comment (klueska) [reply pkg/nvml/api.go:636]: Not at the moment. It was fairly easy to do though. I just grepped for `func (l* library)` copied that in and did some search and replace on it.

File: pkg/nvml/lib.go (hunk 1 above)
Comment (klueska) [reply pkg/nvml/lib.go:152]: The nvml "functions" in this block come from #defines in nvml.h.  The underlying versioned functions that get assigned to these variables are generated as part of the bindings with the `v1`, `v2`, `v3`, etc. prefixes and then assigned in `updateVersionedSymbols` below.

The difference with these functions is that we need to wrap the underlying `nvml*_v1`, `nvml*_v2`, etc. functions with some extra, version specific logic. Because of this, we create a level of indirection before assinging them back out to the "unversioned" variable.

File: pkg/nvml/gpm.go (hunk 3 above)
Comment (klueska) [reply pkg/nvml/gpm.go:69]: No, you don't need to. If you just want the latest, you call e.g. `GpmQueryDeviceSupport()` directly. Only if you care about a specific version do you call `GpmQueryDeviceSupportV().V1()`, `GpmQueryDeviceSupportV().V2()`, etc.

File: pkg/nvml/api.go (hunk 4 above)
Comment (klueska) [reply pkg/nvml/api.go:575]: I am doing a pass now to lowercase all of the variable names. It was causing problems with `moq` as well. It's the reason this is still a draft PR until I get these updated.

File: pkg/nvml/api.go (hunk 5 above)
Comment (klueska) [reply pkg/nvml/api.go:282]: Yes, it is a bug. However, if I do it as a separate PR, then I can't make the updates in the commit that follows to turn ComputeInstance into an Interface. I'd rather have all the interface conversion in a single PR. I could look at maybe updating it in a "pre" PR, but I don't know how hard it will be to rebase if I do that.

File: pkg/nvml/api.go (hunk 6 above)
Comment (klueska) [reply pkg/nvml/api.go:295]: Ditto here, I will look into it as a "pre" PR.

File: pkg/nvml/api.go (hunk 7 above)
Comment (klueska) [reply pkg/nvml/api.go:582]: Yes, I am doing a pass now to lowercase all variable names.

File: pkg/nvml/api.go (hunk 6 above)
Comment (elezar) [reply pkg/nvml/api.go:295]: It's not a blocker -- as long as we actually call it out in the release notes.

File: pkg/nvml/api.go (hunk 7 above)
Comment (elezar) [reply pkg/nvml/api.go:582]: We can do it as a follow-up if that gets too much.

Comment (klueska) [reply pkg/nvml/api.go]: I have made this change, updated, and moved the PR out of draft.

**Comment by klueska** [comment]
> Initial change look good. The main question I have is whether to top-level library functions that operate ON other types make sense, or whether we should always use the functions off the types?

I'd be fine returning concrete types for everything (including the top level library) and moving the API definitions to the `testing` package. I went back and forth on this and don't really have a preference, so long as we have a way to pull in a mock version of the library (and its other types) from somewhere.

**Comment by elezar** [comment]
> > Initial change look good. The main question I have is whether to top-level library functions that operate ON other types make sense, or whether we should always use the functions off the types?

No, that's not what I meant -- although it is a valid question. I meant that we have, as an example, three functions that do the same thing:
```
DeviceGetComputeMode(device Device)  (ComputeMode, Return)
library.DeviceGetComputeMode(device Device) (ComputeMode, Return)
Device.GetComputeMode()  (ComputeMode, Return)
```
My question was whether we would consider moving towards requiring the use of `Device.GetComputeMode()` and at least deprecate the other usages? This would drastically reduce the size of the top-level `Interface`.

**Comment by klueska** [comment]
> > > Initial change look good. The main question I have is whether to top-level library functions that operate ON other types make sense, or whether we should always use the functions off the types?
> 
> No, that's not what I meant -- although it is a valid question. I meant that we have, as an example, three functions that do the same thing:
> 
> ```
> DeviceGetComputeMode(device Device)  (ComputeMode, Return)
> library.DeviceGetComputeMode(device Device) (ComputeMode, Return)
> Device.GetComputeMode()  (ComputeMode, Return)
> ```
> 
> My question was whether we would consider moving towards requiring the use of `Device.GetComputeMode()` and at least deprecate the other usages? This would drastically reduce the size of the top-level `Interface`.

I would be inclined to keep them since that is what matches the actual NVML API. That said, we should add tooling to autogenerate the top level Interface as well as the package level aliases from the functions hanging off the `library` type.

File: pkg/nvml/api.go (hunk 2 above)
Comment (elezar) [reply pkg/nvml/api.go]: OK. Ordering the same as `nvml.h` is good -- this should then aid automation and also make diffs easy enough to read. No need for this follow up then.

File: pkg/nvml/api.go (hunk 5 above)
Comment (elezar) [reply pkg/nvml/api.go]: I'm happy to leave this in this PR as long as we remember to call it out in the release notes once we tag a new version.

File: gen/nvml/generateapi.go [hunk 10]
```diff
@@ -0,0 +1,388 @@
+package main
```
Comment (elezar) [inline gen/nvml/generateapi.go:1]: License?

File: pkg/nvml/zz_generated.api.go [hunk 11]
```diff
@@ -0,0 +1,885 @@
+// Generated Code; DO NOT EDIT.
```
Comment (elezar) [inline pkg/nvml/zz_generated.api.go:1]: Do we still want to add the license boilerplate?

File: gen/nvml/generateapi.go [hunk 12]
```diff
@@ -0,0 +1,388 @@
...
+		fmt.Fprintf(writer, output)
+
+		if i < (len(GeneratableInterfaces) - 1) {
+			fmt.Fprintf(writer, "\n")
+		}
+	}
+}
+
+func getWriter(outputFile string) (io.Writer, func() error, error) {
+	if outputFile == "-" {
```
Comment (elezar) [inline gen/nvml/generateapi.go:132]: Is `-` not used for `STDIN`? and `""` for `STDOUT`? (this is a nit).

File: gen/nvml/generateapi.go [hunk 13]
```diff
@@ -0,0 +1,388 @@
...
+		return os.Stdout, nil, nil
```
Comment (elezar) [inline gen/nvml/generateapi.go:133]: Instead of `nil` as the "Closer" can we return:
```
func() error { return nil }
```
then we don't need to check for nil at the close site?

Alternatively, we can return a `io.WriteCloser` and implement a `noopCloser` that we combine with os.Stdout:
```
type noopCloser struct {
   io.Writer
}

func (n *noopCloser) Close() error {
   return nil
}
```
And then our return statement becomes:
```
return &noopCloser{Writer: os.Stdout}, nil
```
(and we can just `return os.Create(outputFile)` directly in the file case).

File: gen/nvml/generateapi.go [hunk 14]
```diff
@@ -0,0 +1,388 @@
...
+		if err != nil {
+			return err
+		}
+
+		gofiles[path] = content
+
+		return nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("walking the embed.FS path: %w\n", err)
```
Comment (elezar) [inline gen/nvml/generateapi.go:243]: embed.FS?

File: gen/nvml/generateapi.go [hunk 15]
```diff
@@ -0,0 +1,388 @@
...
+		}
+
+		comment, err := generateInterfaceComment(p)
+		if err != nil {
+			fmt.Printf("Error: %v", err)
+			return
+		}
+		fmt.Fprintf(writer, comment)
+
+		output, err := generateInterface(*sourceDir, p)
```
Comment (elezar) [inline gen/nvml/generateapi.go:118]: Do we fmt the output automaticallty?

**Review by elezar — COMMENTED** [review]
Some minor comments on the generation toolking to start with.

File: gen/nvml/generateapi.go (hunk 12 above)
Comment (klueska) [reply gen/nvml/generateapi.go:132]: Yes, you are right -- will fix

File: gen/nvml/generateapi.go (hunk 13 above)
Comment (klueska) [reply gen/nvml/generateapi.go:133]: I went with the first suggestion

Comment (klueska) [reply gen/nvml/generateapi.go:133]: As you know, I am not a fan of "noop code" and prefer to be explicit (which is why I hesitated even changing this at all, but this is minor, so its fine.

File: gen/nvml/generateapi.go (hunk 14 above)
Comment (klueska) [reply gen/nvml/generateapi.go:243]: Leftover from an old iteration, removing

File: gen/nvml/generateapi.go (hunk 15 above)
Comment (klueska) [reply gen/nvml/generateapi.go:118]: I now call `make fmt` as the last step in `make bindings` where the call to run this script is invoked.

File: pkg/nvml/api.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (klueska) [reply pkg/nvml/api.go]: The tool has been added in the latest revision.

File: gen/nvml/generateapi.go (hunk 12 above)
Comment (klueska) [reply gen/nvml/generateapi.go]: Done

File: gen/nvml/generateapi.go (hunk 14 above)
Comment (klueska) [reply gen/nvml/generateapi.go]: Done

File: pkg/nvml/zz_generated.api.go (hunk 11 above)
Comment (klueska) [reply pkg/nvml/zz_generated.api.go:17]: Added in latest

File: gen/nvml/generateapi.go (hunk 10 above)
Comment (klueska) [reply gen/nvml/generateapi.go:17]: Added in latest

File: go.mod [hunk 16]
```diff
@@ -2,7 +2,10 @@ module github.com/NVIDIA/go-nvml
 
 go 1.20
 
-require github.com/stretchr/testify v1.9.0
+require (
+	github.com/google/uuid v1.6.0
```
Comment (elezar) [inline go.mod:6]: As a matter of interest, where is this dependency from?

File: gen/nvml/generateapi.go [hunk 17]
```diff
@@ -0,0 +1,420 @@
...
+		return nil, nil, err
+	}
+
+	return file, file.Close, nil
+}
+
+func generateHeader() (string, error) {
+	lines := []string{
+		"/**",
+		"# Copyright 2024 NVIDIA CORPORATION",
```
Comment (elezar) [inline gen/nvml/generateapi.go:161]: As a follow up we may want to read a `boilerplate.txt` from disk instead so as to not have to update the date.

File: pkg/nvml/mock/dgxa100.go
Comment (elezar) [inline pkg/nvml/mock/dgxa100.go:1]: As a thought. Does it make sense to have the "implementations" under something like `mock/system` or `mock/platform`?

**Review by elezar — DISMISSED** [review]
Minor comment about where we put our Mock sytems implementations.

File: go.mod (hunk 16 above)
Comment (klueska) [reply go.mod:6]: It's to generate a UUID for the GPUs in the mock dgxa100 server

File: pkg/nvml/mock/dgxa100.go
Comment (klueska) [reply pkg/nvml/mock/dgxa100.go:1]: I'd be happy with a `dgxa100` folder though. Then we could change the names to `dgxa100.Server` and `dgxa100.Device`, `dgxa100.GpuInstance`, etc.

Comment (klueska) [reply pkg/nvml/mock/dgxa100.go:1]: I made this change -- let me know what you think

File: go.mod (hunk 16 above)
Comment (elezar) [reply go.mod:6]: Ah. OK. Thanks.

File: pkg/nvml/mock/dgxa100.go
Comment (elezar) [reply pkg/nvml/mock/dgxa100.go:1]: > What value does the extra level of nesting bring? 
It's more about organisation than anything else. The mocks themselves are all generated wherease the systems are implemented.

File: pkg/nvml/mock/dgxa100/dgxa100.go [hunk 18]
```diff
@@ -0,0 +1,565 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.  All rights reserved.
```
Comment (elezar) [inline pkg/nvml/mock/dgxa100/dgxa100.go:2]: Nit: The date.

**Review by elezar — DISMISSED** [review]
Thanks for this @klueska.

I have some local changes in the device plugin that could benefit from the "Mock" implementations already!

**Review by elezar — DISMISSED** [review]

File: pkg/nvml/lib.go [hunk 19]
```diff
@@ -85,6 +113,10 @@ func (l *library) load() (rerr error) {
 		return fmt.Errorf("error opening %s: %w", l.path, err)
 	}
 
+	// Update the errorStringFunc to point to nvml.ErrorString
+	errorStringFunc = nvmlErrorString
```
Comment (elezar) [inline pkg/nvml/lib.go:117]: Do we need to set this to the default once we close the library?

File: pkg/nvml/return.go [hunk 20]
```diff
@@ -19,17 +19,27 @@ import (
 )
 
 // nvml.ErrorString()
-func ErrorString(r Return) string {
-	if err := GetLibrary().Lookup("nvmlErrorString"); err != nil {
-		return fallbackErrorStringFunc(r)
-	}
-	return nvmlErrorString(r)
+func (l *library) ErrorString(r Return) string {
+	return r.Error()
```
Comment (elezar) [inline pkg/nvml/return.go:23]: We either need to keep the lookup logic as before or reset the default if the library is unloaded.

**Review by elezar — COMMENTED** [review]
Thanks. Looks good.

Minor comment regarding the error string.

File: pkg/nvml/lib.go [hunk 21]
```diff
@@ -43,27 +53,45 @@ type library struct {
...
+	}
+	if o.flags == 0 {
+		o.flags = defaultNvmlLibraryLoadFlags
+	}
+
+	l.path = o.path
+	l.dl = dl.New(o.path, o.flags)
+}
+
+func (l *library) GetExtendedInterface() ExtendedInterface {
```
Comment (elezar) [inline pkg/nvml/lib.go:88]: I'm not sold on `GetExtendedInterface` and would prefer `GetExtensions()`, but this is not a blocker. I'm happy to keep it as is and revisit this later once we have some usage examples.

File: pkg/nvml/return.go (hunk 20 above)
Comment (klueska) [reply pkg/nvml/return.go:23]: I can't leave it as before because in the r.Error() and r.String() methods I don't have access to `l`. I will update it reset the function on unload.

File: pkg/nvml/lib.go (hunk 21 above)
Comment (klueska) [reply pkg/nvml/lib.go:88]: Updated as suggested.

Comment (klueska) [reply pkg/nvml/lib.go]: Actually, I went for a hybrid -- the interface is called `ExtendedInterface` to be symmetrical with `Interface`, but the method call is `Extensions()` (which is the only actual symbol anyone will see in practice), e.g.:
```
nvml.Extensions().LookupSymbol("symbol")
```
or
```
envml := nvml.Extensions()
envml.LookupSymbol("symbol")
```

Its possible that the type itself might be embedded in a struct, but I only see that happening in testing (if at all) and I like the symmetry of that with `Interface` for the core API.

File: pkg/nvml/return.go (hunk 20 above)
Comment (klueska) [reply pkg/nvml/return.go:23]: Updated.

File: pkg/nvml/lib.go (hunk 19 above)
Comment (klueska) [reply pkg/nvml/lib.go:117]: Yes. Fixed.

File: pkg/nvml/lib.go (hunk 21 above)
Comment (elezar) [reply pkg/nvml/lib.go]: That sounds fine. I'm not too concerned about the type name, since as you mention that isn't visible at the call site.

**Review by elezar — APPROVED** [review]
Looks good now!
